src/lib/profiles/weave-tunneling/WeaveTunnelAgent.cpp

```diff
@@ -934,6 +939,150 @@ void WeaveTunnelAgent::ParseDestinationIPAddress(const PacketBuffer &inMsg, IPAd
 #endif // !WEAVE_SYSTEM_CONFIG_USE_LWIP
 }
 
+#if WEAVE_CONFIG_ENABLE_MESSAGE_CAPTURE
+
+struct PseudoHeader {
+    uint16_t    PayloadLength;
+    uint16_t    Protocol;
+    uint8_t     SrcAddr[16];
+    uint8_t     DstAddr[16];
+};
+
+uint32_t WeaveTunnelAgent::Checksum(uint16_t *buf, uint16_t len)
+{
+    uint32_t sum    = 0;
+    uint16_t result = 0;
+
+    while (len > 1)
+    {
+        sum += *buf++;
+        len -= 2;
+    }
+
+    if (len == 1)
+    {
+        *(uint8_t *)(&result) = *(uint8_t *)buf;
+        sum += result;
+    }
+
+    return sum;
+}
+
+WEAVE_ERROR WeaveTunnelAgent::ComputeUDPChecksumForIPv6Pkt(PacketBuffer *inMsg, const IPAddress &srcAddr, const IPAddress &destAddr)
+{
+    WEAVE_ERROR       err   = WEAVE_NO_ERROR;
+    uint8_t           *p    = NULL;
+    uint16_t          *q    = NULL;
+    PseudoHeader  payloadWithPseudoHdr;
+    uint16_t udpPktLen = 0;
+    uint32_t sum    = 0;
+    uint16_t *currCSumPtr = 0;
+    uint16_t cSumDataLen = 0;
+
+    // The beginning of the IPv6 packet
+    p = inMsg->Start() + TUN_HDR_SIZE_IN_BYTES;
+
+    // The paylload of the IPv6 packet
+    q = (uint16_t *)(p + sizeof(struct ip6_hdr));
+
+#if WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+    struct udphdr *udpHdr  = (struct udphdr *)(q);
+
+    // Initialize checksum to 0
+    udpHdr->check = 0;
+#else // !WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+    struct udp_hdr *udpHdr =  (struct udp_hdr *)(q);
+
+    // Initialize checksum to 0
+    udpHdr->chksum = 0;
+#endif // !WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+    struct ip6_hdr *ip6hdr = (struct ip6_hdr *)p;
+
+    // Get the UDP length or the IPv6 payload length
+    udpPktLen = inMsg->DataLength() - TUN_HDR_SIZE_IN_BYTES - sizeof(struct ip6_hdr);
+
+    //Fill up IPv6 fields belonging to the pseudo header necessary to calculate the checksum.
+    payloadWithPseudoHdr.PayloadLength   = nl::Weave::Encoding::BigEndian::HostSwap16(udpPktLen);
+    payloadWithPseudoHdr.Protocol      = nl::Weave::Encoding::BigEndian::HostSwap16(NL_PROTO_TYPE_UDP);
+    memcpy(payloadWithPseudoHdr.SrcAddr, &srcAddr, sizeof(payloadWithPseudoHdr.SrcAddr));
+    memcpy(payloadWithPseudoHdr.DstAddr, &destAddr, sizeof(payloadWithPseudoHdr.DstAddr));
+
+    // Calculate checksum for the pseudo header
+    sum = Checksum((uint16_t*)&payloadWithPseudoHdr, sizeof(PseudoHeader));
+    // Calculate the UDP Datagram checksum and add to the sum.
+    sum += Checksum(q, udpPktLen);
+
+    // Adjust for carryover
+    sum = (sum >> 16) + (sum & 0xffff);
+    sum += (sum >> 16);
+    sum = ~sum;
```

Comment:
Lines 1011-1019 are a classic UDP checksum / RFC 768.  Is there a method we could call instead?

Reply:
Alternatively this could be broken out as a separate small function.